src/pitaya/src/ffi/metrics.rs

```diff
@@ -0,0 +1,239 @@
+use async_trait::async_trait;
+use std::{
+    ffi::{c_void, CString},
+    os::raw::c_char,
+};
+
+#[repr(C)]
+pub struct PitayaMetricsOpts {
+    pub namespace: *const c_char,
+    pub subsystem: *const c_char,
+    pub name: *const c_char,
+    pub help: *const c_char,
+    pub variable_labels: *mut *mut c_char,
+    pub variable_labels_count: u32,
+    pub buckets: *mut f64,
+    pub buckets_count: u32,
+}
+
+type PitayaRegisterFn = extern "C" fn(user_data: *mut c_void, opts: PitayaMetricsOpts);
+
+type PitayaIncCounterFn = extern "C" fn(
+    user_date: *mut c_void,
+    name: *const c_char,
+    labels: *mut *const c_char,
+    labels_count: u32,
+);
+
+type PitayaObserveHistFn = extern "C" fn(
+    user_data: *mut c_void,
+    name: *const c_char,
+    value: f64,
+    labels: *mut *const c_char,
+    labels_count: u32,
+);
+
+type PitayaSetGaugeFn = extern "C" fn(
+    user_data: *mut c_void,
+    name: *const c_char,
+    value: f64,
+    labels: *mut *const c_char,
+    labels_count: u32,
+);
+
+type PitayaAddGaugeFn = extern "C" fn(
+    user_data: *mut c_void,
+    name: *const c_char,
+    value: f64,
+    labels: *mut *const c_char,
+    labels_count: u32,
+);
+
+pub struct PitayaMetricsReporter {
+    register_counter_fn: PitayaRegisterFn,
+    register_histogram_fn: PitayaRegisterFn,
+    register_gauge_fn: PitayaRegisterFn,
+    inc_counter_fn: PitayaIncCounterFn,
+    observe_hist_fn: PitayaObserveHistFn,
+    set_gauge_fn: PitayaSetGaugeFn,
+    add_gauge_fn: PitayaAddGaugeFn,
+    user_data: super::PitayaUserData,
+}
+
+fn generic_register(
+    register_fn: extern "C" fn(*mut c_void, PitayaMetricsOpts),
+    user_data: *mut c_void,
+    mut opts: crate::metrics::Opts,
+) {
+    // TODO(lhahn): this function allocates a lot of unnecessary memory.
+    // consider improving this in the future.
+
+    let mut variable_labels_ptr: Vec<*mut c_char> = opts
+        .variable_labels
+        .iter_mut()
+        .map(|s| s.as_mut_ptr() as *mut c_char)
```

Comment:
We need to allocate `CString`s for these strings too 😞

Reply:
Damn it, true :/. I'll change that.